File: Arduino software/MS5840_multiple_sensors/src/MS5840.cpp
#include "MS5840.h"
#include "math.h"
/**
* \brief Class constructor
*
*/
MS5840::MS5840(int address) {this->address=address;}

uint8_t MS5840::reset(){
    uint8_t res= sendCommand(CMD_RESET);
	return res;
    // delayMicroseconds(9);
}

uint8_t MS5840::read_coefficient()
// Initialize library for subsequent pressure measurements
{
	uint8_t i;
	uint8_t highByte,lowByte;
	for(i = 0; i <= 7; i++)
  {
		sendCommand(CMD_PROM + (i * 2));
		Wire.requestFrom( this->address, 2);
		
		while(Wire.available()){
			highByte = Wire.read();
			lowByte = Wire.read();
		}
		
		this->coefficient[i] = (highByte << 8)|lowByte;
	// Uncomment below for debugging output.
	//	Serial.print("C");
	//	Serial.print(i);
	//	Serial.print("= ");
	//	Serial.println(coefficient[i]);
	}

	return 0;
}

void MS5840::sendADCconversion(measurement _measurement, precision _precision)
{
	sendCommand(CMD_ADC_CONV + _measurement + _precision);
}

uint32_t MS5840::getADCconversion(measurement _measurement)
{
	sendCommand(CMD_ADC_READ);
    Wire.requestFrom(this->address, 3);
	uint8_t highByte = 0, midByte = 0, lowByte = 0;
    while(Wire.available())
	{
		highByte = Wire.read();
		midByte = Wire.read();
		lowByte = Wire.read();
	}
    
	switch( _measurement )
	{
		case PRESSURE : this->D1= ((uint32_t)highByte << 16) + ((uint32_t)midByte << 8) + lowByte; return this->D1;
		case TEMPERATURE : this->D2= ((uint32_t)highByte << 16) + ((uint32_t)midByte << 8) + lowByte; return this->D2;
	}

	
}

uint8_t MS5840::sendCommand(uint8_t command)
{
	Wire.beginTransmission( this->address);
	Wire.write(command);
	uint8_t i2c_status=Wire.endTransmission();
	return i2c_status;
}



void MS5840::process_data()

{
	//Create Variables for calculations
	int32_t temp_calc;
	int32_t pressure_calc;

	int32_t dT;

	//Now that we have a raw temperature, let's compute our actual.
	dT = this->D2 - ((int32_t)coefficient[5] << 8);
	temp_calc = (((int64_t)dT * coefficient[6]) >> 23) + 2000;

	// TODO TESTING  _temperature_actual = temp_calc;

	//Now we have our first order Temperature, let's calculate the second order.
	int64_t TI, OFFI, SENSI, OFF, SENS, OFF2, SENS2; //working variables

	if (temp_calc > 2000)
	// If temp_calc is above 20.0C
	{
		TI = 0;
		OFFI = 0;
		SENSI = 0;

    }
	else
	// If temp_calc is below 20.0C
	{
        if (temp_calc > 1000)
        // If temp_calc is above 10.0C
        {
            TI = 12 * ((uint64_t)dT * dT)/pow(2,35);
    		OFFI = 30 * ((temp_calc - 2000) * (temp_calc - 2000)) / 256;
    		SENSI = 0;
        }
        else
        // If temp_calc is below 10.0C
        {
            TI = 14 * ((uint64_t)dT * dT)/pow(2,35);
    		OFFI = 35 * ((temp_calc - 2000) * (temp_calc - 2000)) / 8;
    		SENSI = 63 * ((temp_calc - 2000) * (temp_calc - 2000)) / 32;
        }

	}

	// Now bring it all together to apply offsets

	OFF = ((int64_t)coefficient[2] << 17) + (((coefficient[4] * (int64_t)dT)) >> 6);
	SENS = ((int64_t)coefficient[1] << 16) + (((coefficient[3] * (int64_t)dT)) >> 7);

	temp_calc = temp_calc - TI;
	OFF2 = OFF - OFFI;
	SENS2 = SENS - SENSI;

	// Now lets calculate the pressure


	pressure_calc = (((SENS2 * this->D1) / 2097152 ) - OFF2) / 32768;

	this->_temperature_actual = temp_calc ;
	this->_pressure_actual = pressure_calc ; // 10;// pressure_calc;


}

int32_t MS5840::get_temprature(){
	return _temperature_actual;
}


int32_t MS5840::get_pressure(){
	return _pressure_actual;
}

uint16_t * MS5840::get_coefficient(){
	return coefficient;
}

File: Arduino software/scripts/MS5840.py
import numpy as np
class MS5840():
    def __init__(self):
        self.coefficient=np.zeros(8,dtype=np.uint16)
        self.D1_list=[]
        self.D2_list=[]
        self.pressure_time_list=[]
        self.temperature_time_list=[]
        self.pressure_list=[]
        self.temperature_list=[]
    def read_coefficient(self,coefficient):
        for i in range(8):
            self.coefficient[i]=np.int32(int(coefficient[i]))
    def read_D1(self,t,D1):
        self.D1_list.append(D1)
        self.pressure_time_list.append(t)
    def read_D2(self,t,D2):
        self.D2_list.append(D2)
        self.temperature_time_list.append(t)
    

    def process_data(self):
        if len(self.D2_list) > len(self.D1_list):
            self.D2_list=self.D2_list[:len(self.D1_list)]
            self.temperature_time_list=self.temperature_time_list[:len(self.D1_list)]
        if len(self.D1_list) > len(self.D2_list):
            self.D1_list=self.D1_list[:len(self.D2_list)]
            self.pressure_time_list=self.pressure_time_list[:len(self.D2_list)]
        for i in range(len(self.D1_list)):
            # print(i)
            pressure,temperature=self._process_data(self.D1_list[i],self.D2_list[i])
            self.pressure_list.append(pressure)
            self.temperature_list.append(temperature)

            
    def _process_data(self,D1,D2):
        temp_calc=np.int32(0)
        pressure_calc=np.int32(0)
        dT=np.int32(0)
        # print(D2)
        # print(type(D2))
        # print((np.int32(self.coefficient[5]) << 5))
        # print(D2)
        # print(self.coefficient)
        dT=D2-(np.int32(self.coefficient[5]) << 8)
        # print(dT)
        # print(coefficient[6])
        
        temp_calc=np.int32((np.int64(dT)*np.int64(self.coefficient[6]) >> 23) +2000)
        # print(temp_calc)
        TI=np.int64(0)
        OFFI=np.int64(0)
        SENSI=np.int64(0)
        OFF=np.int64(0)
        SENS=np.int64(0)
        OFF2=np.int64(0)
        SENS2=np.int64(0)
        if temp_calc > 2000:
            TI=0
            OFFI=0
            SENSI=0
        else:
            if temp_calc > 1000:
               TI = np.int64(12 * np.int64(dT*dT)/ np.power(2,35))
               OFFI = np.int64(30 * ((temp_calc - 2000) * (temp_calc - 2000)) / 256)
               SENSI = 0
            else:
                TI = np.int64(14 * (np.int64(dT*dT))/np.power(2,35))
                OFFI = np.int64(35 * ((temp_calc - 2000) * (temp_calc - 2000)) / 8)
                SENSI =  np.int64(63 * ((temp_calc - 2000) * (temp_calc - 2000)) / 32)
        OFF = (np.int64(self.coefficient[2] << 17)) + (((self.coefficient[4] * np.int64(dT))) >> 6)
        SENS = (np.int64(self.coefficient[1]) << 16) + (((self.coefficient[3] * np.int64(dT))) >> 7)
        temp_calc = temp_calc - TI
        OFF2 = OFF - OFFI
        SENS2 = SENS - SENSI
        pressure_calc = (((SENS2 * D1) / 2097152 ) - OFF2) / 32768
        # print(pressure_calc,temp_calc)
        return pressure_calc,temp_calc
if __name__ == "__main__":
    ms5840=MS5840()
    coefficient=['26785', '41335', '41466', '26840' ,'27317','31225' ,'28780' ,'65535'  ]
    ms5840.read_coefficient(coefficient)
    ms5840.read_D1(0,6735203)
    ms5840.read_D2(0,8073757)
    ms5840.process_data()